File: classic puzzle-easy/feature-extraction.py
#https://www.codingame.com/training/easy/feature-extraction
#FEATURE EXTRACTION
#concept: Dot product
import sys
import math

r,c=[int(i) for i in input().split()]
a,w=[],[]
for i in range(r):
    a+=[[int(j) for j in input().split()]]
m,n=[int(i) for i in input().split()]
for i in range(m):
    w+=[[int(j) for j in input().split()]]

for i in range(0,r-m+1):
    t=[]
    for j in range(0,c-n+1):
        s=0
        for k in range(0,m):
            for l in range(0,n):
                s+=a[i+k][j+l]*w[k][l]
        t+=[s]
    print(*t)

File: classic puzzle-easy/feature-extraction.rb
#https://www.codingame.com/training/easy/feature-extraction
#FEATURE EXTRACTION
#concept: Dot product
r,c=gets.split.map &:to_i
a,w=[],[]
r.times{a+=[gets.split.map(&:to_i)]}
m,n=gets.split.map &:to_i
m.times{w+=[gets.split.map(&:to_i)]}
(0..r-m).map{|i|
  x=[]
  (0..c-n).map{|j|
    s=0
    for k in (0...m)
      (0...n).each do |l|; s+=a[i+k][j+l]*w[k][l]; end
    end
    x+=[s]
  }
  puts x*" "
}
